### DISCUSSION TITLE
Processing multiple datasets from using a single Data class

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#4769] [Q&A] [answered]
Hi Matt, I'm trying to process many raw files ```mydata_1.npz```, ```mydata_2.npz```, ... ,```mydata_100.npz``` located in the raw folder.
I made a Dataset class called ```MyData```.

```Python
class MyData(InMemoryDataset):
    def __init__(self, index, root, transform=None):
        super(MyData, self).__init__(index, root, transform)
        self.data, self.slices = torch.load(self.processed_paths[0])
        self.index = index

    @property
    def raw_file_names(self):
        return "mydata_{}.npz".format(self.index)

    @property
    def processed_file_names(self):
        return "data_{}.pt".format(self.index)

    def process(self):
        raw_data = np.load(self.raw_paths[0])
        x = torch.from_numpy(raw_data['x'])
        y = torch.from_numpy(raw_data['y'])
        
        pos = torch.stack([x,y], dim=-1)
        value = torch.from_numpy(raw_data['value'])

        data_list = []
        for i in range(cp.size(0)):
            data = Data(x=value[i].view(-1,1),pos=pos.view(-1,2))
            data_list.append(data)

        torch.save(self.collate(data_list), self.processed_paths[0])
```

However, I don't know how to call multiple data objects at once like we usually do. 

``` Python
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
transform = T.Cartesian(cat=False)
dataset_1 = MyData(index=1, root='./', transform=transform)
dataset_2 = MyData(index=2, root='./', transform=transform)
...
dataset_100 = MyData(index=100, root='./', transform=transform)
```
I don't think we can use ```for``` loop for that though. 
Also, is ```MyData``` class correct for processing multiple raw files?

**Comment by rusty1s** [comment, marked as answer]
Please note that a `Dataset` already groups all your graphs into one place, such that you can do
```python
dataset = MyDataset(...)
data = dataset[0]  # Access first graph
...
```
For this, the `InMemoryDataset` implementation slightly changes:
```python
class MyDataset(InMemoryDataset):
    def __init__(self, root, transform=None):
        super(MyData, self).__init__(root, transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return [f"mydata_{i + 1}.npz" for i in range(100)]

    @property
    def processed_file_names(self):
        return "data.pt"

    def process(self):
        data_list = []
        for path in self.raw_paths:
            raw_data = np.load(path)
            x = torch.from_numpy(raw_data['x'])
            y = torch.from_numpy(raw_data['y'])
            data = Data(x=x, y=y, ...)
            data_list.append(data)
        
        torch.save(self.collate(data_list), self.processed_paths[0])
```

**Reply by hkim716** [reply]
Thanks Matt, it worked.
However, in that case, processed ```data.pt``` is too large in size. 
How could I process ```mydata_{i+1}.npz``` into ```data_{i+1}.pt``` separately?
For example, ```mydata_1.npz``` corresponds to ```data_1.pt```,
```mydata_2.npz``` corresponds to ```data_2.pt```, 
...
```mydata_100.npz``` corresponds to ```data_100.pt```.

I tried 
``` Python
    @property
    def raw_file_names(self):
        return [f"mydata_{i + 1}.npz" for i in range(100)]
    @property
    def processed_file_names(self):
        return [f"data_{i+1}.pt" for i in range(100)]
    def process(self):
        idx=0
        for path in self.raw_paths:
            data_list = []
            raw_data = np.load(path)
            x = torch.from_numpy(raw_data['x'])
            y = torch.from_numpy(raw_data['y'])
            data = Data(x=x, y=y, ...)
            data_list.append(data)
            idx+=1
            torch.save(self.collate(data_list), self.processed_paths[0])
```
but it just created only ```data_1.pt``` in the processed folder.

**Reply by rusty1s** [reply]
This is described here: https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html#creating-larger-datasets
Importantly, you no longer inherit from `InMemoryDataset`.

**Reply by hkim716** [reply]
Thanks Matt.
It helped to generate separate ```data_1.pt```, ```data_2.pt```, ... ,```data_100.pt``` in the processed folder.
I write ```MyData``` class like this with ```len()``` and ```get()```.
However, each ```data_1.pt``` is actually supposed to have 140 graphs that have different ```node feature``` and ```labels```.


```Python
class MyData(Dataset):
    def __init__(self, root, transform=None):
        super(MyData, self).__init__(root, transform)
        
    @property
    def raw_file_names(self):
        return [f"mydata_{i+1}.npz" for i in range(100)]

    @property
    def processed_file_names(self):
        return [f"data_{i+1}.pt" for i in range(100)]
    
    def process(self):
        idx = 0
        for path in self.raw_paths:
            data_list=[]
            raw_data = np.load(path)
            x = torch.from_numpy(raw_data['x'])
            y = torch.from_numpy(raw_data['y'])
            
            pos = torch.stack([x,y,z], dim=-1)
            value = torch.from_numpy(raw_data['value'])
            label = torch.from_numpy(raw_data['conditions'])
            
            edge_index = torch.from_numpy(raw_data['edge_index'])
            edge_index -= 1
            edge_index = edge_index.t().contiguous()
            edge_index = to_undirected(edge_index)

            for i in range(value.size(0)):
                data = Data(x=value[i].view(-1,1),pos=pos.view(-1,2), edge_index=edge_index, y=label[i,:].view(-1))
                data_list.append(data)

            torch.save(data_list, osp.join(self.processed_dir, f'data_{idx}.pt'))
            idx+=1
            
    def len(self):
        return len(self.processed_file_names)
    
    def get(self,idx):
        data = torch.load(osp.join(self.processed_dir, f'data_{idx}.pt'))
        return data
```

When I printed out 
```dataset= MyData(root='./', transform=transform)
print(dataset)
``` 
I see ```MyData(100)```.
but when I tried to ```print(dataset[0])``` to see the first set of 140 data, there is an error 
``` Python
197 
    198             data = self.get(self.indices()[idx])
--> 199             data = data if self.transform is None else self.transform(data)
    200             return data
    201 

~/miniconda3/envs/pyg_172/lib/python3.7/site-packages/torch_geometric/transforms/cartesian.py in __call__(self, data)
     24 
     25     def __call__(self, data):
---> 26         (row, col), pos, pseudo = data.edge_index, data.pos, data.edge_attr
     27 
     28         cart = pos[row] - pos[col]

AttributeError: 'list' object has no attribute 'edge_index'
```

I think I need to fix ```get()```, but I'm not sure how to do it. 
What I'm trying to get is when I ```print(dataset[0])```, I want to see a data list that contains 140 graph dataset like 
```[Data(x=[5581, 1], edge_index=[2, 32836], y=[3], pos=[5581, 3], cluster=[6]), Data(x=[5581, 1], edge_index=[2, 32836], y=[3], pos=[5581, 3], cluster=[6]), Data(x=[5581, 1], edge_index=[2, 32836], y=[3], pos=[5581, 3], cluster=[6]), ............ Data(x=[5581, 1], edge_index=[2, 32836], y=[3], pos=[5581, 3], cluster=[6])]```, so I can pick one of these using ```dataset[0][129]```, which is 130th graph data. 

How could I correctly (and efficiently) access to ```data_n.pt``` and the specific graph in that?

**Reply by rusty1s** [reply]
You could try to run
```python
def get(self,idx):
    data_list = torch.load(osp.join(self.processed_dir, f'data_{idx}.pt'))
    return Batch.from_data_list(data_list)
```
Better support for this will be tracked in https://github.com/pyg-team/pytorch_geometric/issues/4543

**Reply by hkim716** [reply]
Thanks Matt, it worked.
I have another question about how we could minimize memory used during training a GNN. 
As you know, my first (poorly designed) data interface was to use multiple ```MyData_n(InMemoryDataset)``` class to load multiple graphs such as 
```Python 
from mydata1 import MyData_1
from mydata2 import MyData_2
...
from mydata100 import MyData_100

dataset_1 = MyData_1(root='./', transform=transform)
dataset_2 = MyData_2(root='./', transform=transform)
...
dataset_100 = MyData_100(root='./', transform=transform)
```
Each ```MyData_n``` generates ```data_n.pt``` in the processed folder, and ```data_n.pt (=dataset_n)``` has 130 graph samples. I know this is very inefficient way to load datasets into a GNN model.
By using this way, I found that a lot of ```cuda``` memory was taken while training the model, which sometimes resulted in OOM error.

As you suggested, now I can use one ```MyData(Dataset)``` class, also generating ```data_n.pt``` in the processed folder.
```Python
dataset= MyData(root='./', transform=transform)
dataset[i][j]
```

I know ```MyData(Dataset)``` is much wiser way to load dataset, but is there any difference in ```cuda``` memory used between the two (```InMemoryDataset``` and ```Dataset```) methods? For example, when I use ```MyData(Dataset)```, could I reduce the ```cuda``` memory required to train the model compared to using multiple ```InMemoryDataset``` classes? 
I think ```cuda``` memory used should be similar in both ways because ```.pt``` files are saved on ```cpu```, 
and as long as 
```Python
train_loader = DataLoader(train_dataset, batch_size=batch_size)
model = gnn_model().to('cuda')
for data in train_loader:
        data = data.to('cuda')
        pred = model(data)
```
this part does not change, the amount of data transferred to ```cuda``` should be the same...

**Reply by rusty1s** [reply]
Yes, `cuda` memory will be similar. How big is one `data` object coming from the `train_loader`? You can either try to decrease the number of elements you save into one `data` chunk. Alternatively, you can look into [this section](https://pytorch-geometric.readthedocs.io/en/latest/notes/sparse_tensor.html), which goes over some tips on how to reduce GPU memory consumption inside PyG.

**Reply by hkim716** [reply]
I found that when I load test dataset, the memory consumption dramatically increases.

``` Python
def train(epoch):
    model.train()
    total_loss = 0
    it = 0
    for data in train_loader:
        it += 1
        data = data.to(device)
        optimizer.zero_grad()
        out = model(data)
        loss = F.mse_loss(out, data.x)
        total_loss += loss
        loss.backward()
        optimizer.step()

        if it%10==0:
            a = torch.cuda.memory_allocated(0)
            print(f'train memory allocated:{a}')

def test(epoch):
        model.eval()
        total_loss = 0
        it = 0
        for data in test_loader:
        it += 1
        data = data.to(device)
        out = model(data)
        test_loss = F.mse_loss(out, data.x)
        total_loss += test_loss

        if it%10==0:
            print(f'Epoch:{epoch} it:{it} test_loss:{test_loss:.8f}')
            a = torch.cuda.memory_allocated(0)
            print(f'test memory allocated:{a}')

for epoch in range(num_epochs):
    train_loss = train(epoch)
    test_loss = test(epoch)

```

The ratio of train/test dataet is ```95%/5%``` (I made it extreme because I tried to reduce the memory usage during ```test( )```), but the memory usage is too high when ```test_loss = test(epoch)``` is running. 

```
Epoch:0 it:10 train_loss:0.43061042
train memory allocated:373515264
...
Epoch:0 it:10 test_loss:0.19424717
test memory allocated:10391087104
...

Epoch:1 it:10 train_loss:0.16549078
train memory allocated:12544779264
...
Epoch:1 it:10 test_loss:0.12883960
test memory allocated:22567092736
```
And eventually it ends up with OOM error at ```Epoch:1```.
I don't know why the memory usage jumps that much when test data is being loaded ```test_loader = DataLoader(test_dataset,batch_size=batch_size,shuffle=False)```

**Reply by rusty1s** [reply]
You need to run your `test` function with `torch.no_grad()`:
```python
@torch.no_grad()
def test(epoch):
    pass
```
Otherwise, the computation graph is never properly released.

**Reply by hkim716** [reply]
Oh, you are right! I forgot to use that >.<
Thanks!
